File: Week4_Enterprise_Practice/Lesson07_Document_Review_Agent/AgentTeamProject/backend/app/api/review_config.py
from typing import Any, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.errors import APIError
from app.database import get_db
from app.models.session import ReviewSession
from app.services import review_agent_service, review_config_service


router = APIRouter()


class ExecutorTypePayload(BaseModel):
    id: Optional[str] = None
    label: Optional[str] = None
    description: Optional[str] = None
    enabled: Optional[bool] = True


class CheckItemPayload(BaseModel):
    id: Optional[str] = None
    topic_id: Optional[str] = None
    rule_id: Optional[str] = None
    executor_type_id: Optional[str] = None
    review_type: Optional[str] = None
    review_sub_type: Optional[str] = None
    status: Optional[str] = "pending"
    conclusion: Optional[str] = None
    evidence_scope: Optional[dict[str, Any]] = Field(default_factory=dict)
    target_fields: Optional[list[str]] = Field(default_factory=list)
    regulation_clauses: Optional[list[str]] = Field(default_factory=list)
    review_criteria: Optional[str] = None
    expected_result: Optional[str] = None
    failure_conditions: Optional[list[str]] = Field(default_factory=list)
    source_rule_snapshot: Optional[dict[str, Any]] = Field(default_factory=dict)
    expert_brief: Optional[dict[str, Any]] = None
    enabled: Optional[bool] = True


class PreviewCheckItemPayload(CheckItemPayload):
    session_id: str


@router.get("")
def get_review_config() -> dict[str, Any]:
    try:
        return review_config_service.load_review_config()
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc


@router.get("/executor-types")
def list_executor_types() -> dict[str, Any]:
    try:
        return {"items": review_config_service.list_executor_types()}
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc


@router.post("/executor-types")
def create_executor_type(payload: ExecutorTypePayload) -> dict[str, Any]:
    try:
        item = review_config_service.create_executor_type(payload.model_dump(exclude_none=True))
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc
    return item


@router.patch("/executor-types/{executor_id}")
def update_executor_type(executor_id: str, payload: ExecutorTypePayload) -> dict[str, Any]:
    try:
        return review_config_service.update_executor_type(executor_id, payload.model_dump(exclude_unset=True, exclude_none=True))
    except KeyError as exc:
        raise APIError.not_found("ExecutorType") from exc
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc


@router.delete("/executor-types/{executor_id}")
def delete_executor_type(executor_id: str) -> dict[str, Any]:
    try:
        review_config_service.delete_executor_type(executor_id)
    except KeyError as exc:
        raise APIError.not_found("ExecutorType") from exc
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc
    return {"id": executor_id, "deleted": True}


@router.get("/check-items")
def list_check_items(topic_id: Optional[str] = Query(default=None)) -> dict[str, Any]:
    try:
        return {"items": review_config_service.list_check_item_specs(topic_id)}
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc


@router.post("/check-items")
def create_check_item(payload: CheckItemPayload) -> dict[str, Any]:
    try:
        item = review_config_service.create_check_item_spec(payload.model_dump(exclude_unset=True, exclude_none=True))
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc
    return item


@router.post("/check-items/preview")
def preview_check_item(payload: PreviewCheckItemPayload, db: Session = Depends(get_db)) -> dict[str, Any]:
    session = db.query(ReviewSession).filter(ReviewSession.id == payload.session_id).first()
    if not session:
        raise APIError.not_found("ReviewSession")
    try:
        return review_config_service.preview_check_item_spec(
            payload.session_id,
            payload.model_dump(exclude_unset=True, exclude_none=True, exclude={"session_id"}),
            db,
        )
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc
    except review_agent_service.ReviewAgentUnavailable as exc:
        raise APIError.service_unavailable(str(exc)) from exc


@router.patch("/check-items/{item_id}")
def update_check_item(item_id: str, payload: CheckItemPayload) -> dict[str, Any]:
    try:
        return review_config_service.update_check_item_spec(item_id, payload.model_dump(exclude_unset=True, exclude_none=True))
    except KeyError as exc:
        raise APIError.not_found("ReviewCheckItem") from exc
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc


@router.delete("/check-items/{item_id}")
def delete_check_item(item_id: str) -> dict[str, Any]:
    try:
        review_config_service.delete_check_item_spec(item_id)
    except KeyError as exc:
        raise APIError.not_found("ReviewCheckItem") from exc
    except ValueError as exc:
        raise APIError.bad_request(str(exc)) from exc
    return {"id": item_id, "deleted": True}

File: Week4_Enterprise_Practice/Lesson07_Document_Review_Agent/AgentTeamProject/frontend/src/app/api/reviewConfig.ts
import { apiClient } from './client';

export interface ExecutorType {
  id: string;
  label: string;
  description: string;
  enabled: boolean;
}

export interface ExpertReviewBrief {
  item_name: string;
  review_objective: string;
  evidence_instruction: string;
  judgement_basis: string;
  pass_condition: string;
  issue_condition: string;
  regulation_text?: string;
}

export interface CheckItemPayload {
  id?: string;
  topic_id?: string;
  rule_id?: string;
  executor_type_id?: string;
  review_type?: string;
  review_sub_type?: string;
  status?: string;
  conclusion?: string;
  evidence_scope?: Record<string, unknown>;
  target_fields?: string[];
  regulation_clauses?: string[];
  review_criteria?: string;
  expected_result?: string;
  failure_conditions?: string[];
  evidence_slots?: Array<Record<string, unknown>>;
  formula_checks?: Array<Record<string, unknown>>;
  expert_brief?: ExpertReviewBrief;
  source_rule_snapshot?: Record<string, unknown>;
  enabled?: boolean;
}

export interface PreviewCheckItemPayload extends CheckItemPayload {
  session_id: string;
}

export interface EvidenceAnchor {
  page?: number;
  block_id?: string;
  bbox?: number[];
  coordinate_mode?: string;
  page_width?: number | null;
  page_height?: number | null;
}

export interface RetrievalMatch {
  chunk_id?: string;
  page?: number;
  page_end?: number;
  primary_page?: number;
  page_range?: number[];
  chunk_index?: number;
  section?: string;
  anchors?: EvidenceAnchor[];
  block_ids?: string[];
  bbox_count?: number;
  score?: number;
  vector_score?: number;
  bm25_score?: number;
  vector_rank?: number;
  bm25_rank?: number;
  neighbor_rank?: number;
  final_rank?: number;
  retrieval_sources?: string[];
  source_ranks?: Record<string, unknown>;
  rerank_score?: number;
  rerank_rank?: number;
  neighbor_of?: string;
  text?: string;
  [key: string]: unknown;
}

export interface ProjectCompositionSource extends RetrievalMatch {
  material_type?: string;
}

export interface ProjectCompositionFieldComparison {
  field?: string;
  label?: string;
  status?: 'match' | 'mismatch' | 'missing' | string;
  body_value?: number | null;
  reference_value?: number | null;
  difference?: number | null;
}

export interface ProjectCompositionConsistency {
  status?: 'match' | 'mismatch' | 'needs_review' | string;
  reason?: string;
  body_source?: ProjectCompositionSource | null;
  reference_source?: ProjectCompositionSource | null;
  field_comparisons?: ProjectCompositionFieldComparison[];
}

export interface PreviewAgentTrace {
  query?: string;
  retrieval_mode?: string;
  llm_model?: string;
  persisted?: boolean;
  artifact_dir?: string;
  chunk_count?: number;
  facts_available?: boolean;
  cross_chapter_findings_available?: boolean;
  vector_store?: string;
  [key: string]: unknown;
}

export interface EvidenceSlotPackage {
  source?: string;
  slot_count?: number;
  required_slot_count?: number;
  matched_required_slot_count?: number;
  missing_required_slot_ids?: string[];
  truncated?: boolean;
  slot_limit?: number;
  truncated_required_slot_ids?: string[];
  slots?: Array<Record<string, unknown>>;
}

export interface FormulaCheckResults {
  source?: string;
  check_count?: number;
  pass_count?: number;
  fail_count?: number;
  missing_count?: number;
  unsupported_count?: number;
  checks?: Array<Record<string, unknown>>;
}

export interface PreviewCheckItemResponse {
  check_item: CheckItemSpec;
  evidence_bundle: {
    evidence_texts?: string[];
    evidence_locations?: Array<Record<string, unknown>>;
    retrieval_matches?: RetrievalMatch[];
    matched_target_fields?: string[];
    missing_target_fields?: string[];
    structured_facts?: unknown[];
    cross_reference_findings?: unknown[];
    project_composition_consistency?: ProjectCompositionConsistency;
    evidence_slot_package?: EvidenceSlotPackage;
    formula_check_results?: FormulaCheckResults;
    langextract_grounding?: unknown[] | Record<string, unknown>;
    regulation_context?: Array<Record<string, unknown>>;
    retrieval_score?: number;
    source?: string;
    [key: string]: unknown;
  };
  precheck_result: {
    executor_type_id?: string;
    handler_id?: string;
    execution_status?: string;
    checks?: Array<Record<string, unknown>>;
    evidence_scope?: Record<string, string[]>;
    target_fields?: string[];
    regulation_clauses?: string[];
    llm_required?: boolean;
    next_action?: string;
    summary?: string;
    [key: string]: unknown;
  };
  review_conclusion: {
    status?: string;
    summary?: string;
    actual_value?: string;
    expected_value?: string;
    fix_suggestion?: string;
    confidence?: number;
    next_action?: string;
    llm_required?: boolean;
    [key: string]: unknown;
  };
  suggested_rule_improvements: string[];
  agent_trace?: PreviewAgentTrace;
}

export interface CheckItemSpec {
  id: string;
  topic_id: string;
  rule_id: string;
  executor_type_id: string;
  review_type: string;
  review_sub_type: string;
  status: string;
  conclusion: string;
  evidence_scope: Record<string, unknown>;
  target_fields: string[];
  regulation_clauses: string[];
  review_criteria: string;
  expected_result: string;
  failure_conditions: string[];
  evidence_slots: Array<Record<string, unknown>>;
  formula_checks: Array<Record<string, unknown>>;
  source_rule_snapshot: Record<string, unknown>;
  enabled: boolean;
}

export function listExecutorTypes(): Promise<{ items: ExecutorType[] }> {
  return apiClient.get<{ items: ExecutorType[] }>('/review-config/executor-types');
}

export function createCheckItem(body: CheckItemPayload): Promise<CheckItemSpec> {
  return apiClient.post<CheckItemSpec>('/review-config/check-items', body);
}

export function previewCheckItem(body: PreviewCheckItemPayload): Promise<PreviewCheckItemResponse> {
  return apiClient.post<PreviewCheckItemResponse>('/review-config/check-items/preview', body);
}

export function updateCheckItem(itemId: string, body: CheckItemPayload): Promise<CheckItemSpec> {
  return apiClient.patch<CheckItemSpec>(`/review-config/check-items/${itemId}`, body);
}

export function deleteCheckItem(itemId: string): Promise<{ id: string; deleted: boolean }> {
  return apiClient.delete<{ id: string; deleted: boolean }>(`/review-config/check-items/${itemId}`);
}
